Automattic/woocommerce-payments
# Express checkout on the Pay for Order page

Fixes #4220 

#### Changes proposed in this Pull Request

> **Note:**
> WooCommerce 6.6.0 is required to test this PR, as there was no hook to display the button in earlier versions.
> Either use an up-to-date core version, or check out the code from [this PR](https://github.com/woocommerce/woocommerce/pull/33078).


> **Note 2:**
> I could not get ApplePay to work on my machine on a local, tunneled, or sandbox test site. I assume the Payment Request API should already cover everything, as it works in Chrome.

Adds support for payment request buttons (Apple Pay and Google Pay) to the Pay for Order page.

After the initial exploration, it turned out that this does not require a temporary cart and/or checkout object, as we expected. Although the payment request process allows shipping and billing fields to be edited, as we do during standard checkout, we decided to disable those edits when paying for orders. This behavior follows the standard Pay for Order flow, where the only available action is to just pay for the order.

Once the decision was made to only allow payments through the payment request process, the rest of the code got relatively simplified:

- JS changes mostly include some refactors in order to use most of the same functions, which were used on checkout. The only notable change here is [how the process is initialized on the Pay for Order page](https://github.com/Automattic/woocommerce-payments/pull/4264/files#diff-3b09c95dc3e8c6f0dd75f523bc8fccba95d8308a909d8641c5248c92134b2cacR468-R482).
- On the PHP side, all changes are contained within `includes/class-wc-payments-payment-request-button-handler.php`:
    - [`display_pay_for_order_page_html()`](https://github.com/Automattic/woocommerce-payments/pull/4264/files#diff-01da8e7c10c31988f0205e9327a25bcaaf64e6485c6d391912b68dc66d808353R309) loads all line items and totals for the payment request. Here everything is added as a simple line item, as none of those values would be dynamic once the dialog is loaded.
    - [`ajax_pay_for_order()`](https://github.com/Automattic/woocommerce-payments/pull/4264/files#diff-01da8e7c10c31988f0205e9327a25bcaaf64e6485c6d391912b68dc66d808353R1124) handles the payment. The code here is adapted for payment request AJAX calls, but follows the core code for the pay for order page. In the end, with the right POST fields, we just need to use `$gateway->process_payment()` only with the order ID as a parameter.

Here is how the page should look like (note the numbers/total):

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/5311119/178224514-8b992828-83a0-4459-92ad-d839aeff774d.png">

And the same data in the dialog:

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/5311119/178224679-d01b806e-ba59-4f30-a420-30d1f703485d.png">

#### Testing instructions

Create two orders:

1. Complete checkout with an order, open it in `wp-admin`, and change the status to `pending` (Pending payment).
2. Manually go to `wp-admin` / WooCommerce / Orders / Add New and create an order, assigning it to a customer. There you can add all sorts of product, taxes, fees, shipping, etc. Some of those fields would be harder/more time consuming to populate if you were to use settings or extensions.

With the same user logged in, and in a browser which supports express checkout (Chrome/Safari):

1. Open My Account > Orders > Click the `Pay` button.
2. Observe the displayed line items and totals.
3. Click the payment request button.
4. Verify that the dialog contains the same values as the ones on the page.
5. Complete the payment, and make sure that the order is now `processing`, and has all the necessary notes.

I'd also recommend simulating errors in the back-end, in order to make sure that they are properly reaching the front-end.
